dtm0/dtx.c

```diff
@@ -336,68 +343,88 @@ static void dtx_exec_all_ast_cb(struct m0_sm_group *grp, struct m0_sm_ast *ast)
-		M0_LOG(M0_DEBUG, "dtx " DTID0_F "is stable (PMA)",
-		       DTID0_P(&dtx->dd_txd.dtd_id));
-		m0_sm_state_set(&dtx->dd_sm, M0_DDS_STABLE);
+	m0_mutex_lock(&log->dl_lock);
+	rec = m0_be_dtm0_log_find(log, &txd->dtd_id);
+
+	if (rec != NULL) {
+		dtx = &rec->dlr_dtx;
+		m0_dtm0_tx_desc_apply(&dtx->dd_txd, txd);
+		dtx_log_update(dtx);
+
+		if (dtx->dd_sm.sm_state == M0_DDS_EXECUTED_ALL &&
+		    m0_dtm0_tx_desc_state_exists(&dtx->dd_txd,
+						 M0_DTPS_PERSISTENT)) {
+			M0_ASSERT(dtx->dd_nr_executed ==
+				  dtx->dd_txd.dtd_ps.dtp_nr);
+			M0_LOG(M0_DEBUG, "dtx " DTID0_F "is stable (PMA)",
+			       DTID0_P(&dtx->dd_txd.dtd_id));
+			m0_sm_state_set(&dtx->dd_sm, M0_DDS_STABLE);
+		}
 	}
+	m0_mutex_unlock(&log->dl_lock);
 
-out:
 	m0_free(pma);
 	m0_fop_put_lock(fop); /* it was taken in ::m0_dtm0_dtx_post_pmsg */
 	M0_LEAVE();
 }
 
-M0_INTERNAL void m0_dtm0_dtx_pmsg_post(struct m0_dtm0_dtx *dtx,
+M0_INTERNAL void m0_dtm0_dtx_pmsg_post(struct m0_be_dtm0_log *log,
 				       struct m0_fop      *fop)
 {
-	struct m0_dtm0_pmsg_ast *pma;
+	struct m0_dtm0_pmsg_ast      *pma;
+	struct m0_sm_group           *dtx_sm_grp;
+	struct m0_dtm0_log_rec       *rec;
+	struct dtm0_req_fop          *req = m0_fop_data(fop);
+	const struct m0_dtm0_tx_desc *txd = &req->dtr_txr;
 
 	M0_PRE(fop != NULL);
 	M0_PRE(fop->f_opaque != NULL);
 	M0_PRE(m0_fop_data(fop) != NULL);
-	M0_PRE(dtx->dd_sm.sm_grp != &fop->f_item.ri_rmachine->rm_sm_grp);
 
-	M0_ENTRY("dtx=%p, fop=%p", dtx, fop);
+	M0_ENTRY("fop=%p", fop);
+
 	/*
-	 * XXX: it is an uncommon case where f_opaque is used on the server
-	 * side (rather than on the client side) to associate a FOP with
-	 * a user-defined datum. It helps to connect the lifetime of the
-	 * AST that handles the notice with the lifetime of the FOP that
-	 * has delivered the notice.
+	 * This lookup is done to find the SM group that corresponds
+	 * to DTX SM to post AST. We can not use DTX itself here since
+	 * it can be cleared in another thread and the log lock does
+	 * not protect the DTX between current function and AST.
 	 */
-	pma = fop->f_opaque;
-	M0_ASSERT_INFO(M0_IS0(pma), "PMA cannot be re-used");
-	*pma = (struct  m0_dtm0_pmsg_ast) {
-		.p_dtx = dtx,
-		.p_fop = fop,
-		.p_ast = {
-			.sa_cb = dtx_persistent_ast_cb,
-			.sa_datum = pma,
-		},
-	};
-	m0_fop_get(fop); /* it will be put back by ::dtx_persistent_ast_cb */
-	m0_sm_ast_post(dtx->dd_sm.sm_grp, &pma->p_ast);
+	m0_mutex_lock(&log->dl_lock);
+	rec = m0_be_dtm0_log_find(log, &txd->dtd_id);
+	dtx_sm_grp = rec != NULL ? rec->dlr_dtx.dd_sm.sm_grp : NULL;
+	m0_mutex_unlock(&log->dl_lock);
+
+	if (rec != NULL) {
+		/*
+		 * XXX: it is an uncommon case where f_opaque is used on the
+		 * server side (rather than on the client side) to associate
+		 * a FOP with a user-defined datum. It helps to connect the
+		 * lifetime of the AST that handles the notice with the lifetime
+		 * of the FOP that has delivered the notice.
+		 */
+		pma = fop->f_opaque;
```

Comment:
I see that "fop->f_opaque" is allocated in dtm0_fom_create().
Can you make sure it is freed in somewhere?
Can you make sure it is set to NULL when it is freed to avoid accessing NULL pointer?

Or can we avoid using this "f_opaque"?
Also, can we allocate the memory here just before using it?
```
pma = malloc();
assert(pma != NULL);
*pma = (struct  m0_dtm0_pmsg_ast) {
			.p_log = log,
			.p_fop = fop,
			.p_ast = {
				.sa_cb = dtx_persistent_ast_cb,
				.sa_datum = pma,
			},
		};
fop->f_opaque = pma;
...
```

Reply:
>  Can you make sure it is freed in somewhere?
>  Can you make sure it is set to NULL when it is freed to avoid accessing NULL pointer?

Good points, thanks. Yes, we need to do that. We need to free it in dtm0_fom_fini. And we definitely should set it to NULL in the end of dtx_persistent_ast_cb. cc @sergey-shilov 

> Or can we avoid using this "f_opaque"?
>  Also, can we allocate the memory here just before using it?

That's a tough question. Initially, it was written in a way similar to what you have described. The ast was allocated just before it was posted, and it was not even attached to f_opaque. However, during review I got two points: 

1. We have to connect the lifetime of the FOP and the lifetime of the AST (to make it traceable, and to aid debugging). Later on, it indeed helped me with debugging.
2. We have to pre-allocate as much possible to avoid the need to deal with allocation failures in the middle of execution. For example, when we receive an incoming a request, we have to do our best to allocate the resources need to execute it; otherwise, there is not point to handle a request if we don't have enough resources.

These points convinced me to re-work this part, so that PMA allocation was moved to the FOM initialization, and it got connected to the FOP with help of f_opaque.

What do you think about these points?

Reply:
OK. But we need to update the patch to free the pma .